### DISCUSSION TITLE
Enable type inferring of Spatie Laravel Data in responses

### DISCUSSION THREAD

**Discussion by vilhelmjosander** [dedoc/scramble#211] [Ideas]
Hi all,

This package is really awesome, great work! One thing that would make it even greater would be if it would be possible to enable automatic type inferring for responses of Spaties package Laravel Data (https://github.com/spatie/laravel-data) which is very popular within the Laravel community. We usually return the laravel data object directly in controllers, not an API Resource. 

Example:
Data object
```php
class CustomerData extends Data
{
    public function __construct(
        public int $id,
        public ?string $email,
        public ?string $ssn,
        public ?string $firstName,
        public ?string $lastName,
        public ?string $phone,
        public ?CompanyData $company,
    ) {
    }
```
Controller
```php
 public function user(Request $request): UserData
    {
        return UserData::from($request->user());
    }
```

I tried looking into the extensions documentation to build this extension myself but I can't seem to make it work. Any hints or guidance would be welcome and then I could try implementing it myself and make a PR eventually.

Thanks!

**Comment by romalytvynenko** [comment]
Hey @vilhelmjosander 

It is a great idea. 

I plan to build the extension for [spatie/laravel-data](https://github.com/spatie/laravel-data), I'm just not sure when exactly.

Meanwhile, do you have any particular questions about how to make it work?

I believe that you'd need an extension that helps Scramble to infer return type of `::from` static call, and they you'll need an extension that would describe Scramble how the PHP type (`UserData`) looks like in API.

In general would be cool if you share questions you had when you tried to approach building extensions! I'm really looking toward improving extensions docs.

Thank you 🫶

**Comment by slnw** [comment]
bump! we really need that as well.

**Comment by jericdei** [comment]
It would also be great to infer the request body if the data object is injected in the controller instead of the generic `Request` object.

```php
/**
* Duplicate Check.
*
* Check address for duplicates in the system.
*/
public function __invoke(AddressData $data): JsonResponse
{
  //
}
```

**Reply by escamoteur** [reply]
Absolutely, especially for requests that would allow to use ref types in the Api spec instead of having them defined directly in the Endpoints which would make it better usable for client code generation.
@romalytvynenko any chance to speed up this topic by sponsoring the project or directly paying for this feature?

**Reply by romalytvynenko** [reply]
Hey @escamoteur, will work on its support after the next (0.10.0) release!

**Reply by escamoteur** [reply]
@romalytvynenko great. I tried to send you an email via the https://getpistachio.app/ website but it got bounced with unknown address.

**Reply by romalytvynenko** [reply]
@escamoteur please send again to my personal one: litvinenko95@gmail.com

**Reply by escamoteur** [reply]
@romalytvynenko did you receive my email this time?

**Reply by romalytvynenko** [reply]
@escamoteur yep, replied. Sorry for the delay

**Comment by romalytvynenko** [comment]
@vilhelmjosander @slnw @jericdei Added Laravel Data support in 0.11.0 https://scramble.dedoc.co/blog/scrambledrop-scramble-0110 (as a part of Scramble PRO package)
